### PR TITLE
Update secondary indexes in SQL UPDATE

### PR DESCRIPTION
**Pull request by vivekmenezes**

### CODE REVIEW

**Comment by vivekmenezes** [comment]
Will add a unittest with Pete's new framework soon.

File: sql/update.go [hunk 1]
```diff
@@ -67,21 +70,21 @@ func (p *planner) Update(n *parser.Update) (planNode, error) {
...
 	}
 
-	index := tableDesc.PrimaryIndex
-	indexKey := encodeIndexKeyPrefix(tableDesc.ID, index.ID)
+	primaryIndex := tableDesc.PrimaryIndex
+	primaryIndexKeyPrefix := encodeIndexKeyPrefix(tableDesc.ID, primaryIndex.ID)
 
 	// Evaluate all the column value expressions.
-	vals := make([]parser.Expr, 0, 10)
+	vals := make([]parser.Datum, 0, 10)
```
Comment (tamird) [inline sql/update.go]: what's this change about?

Comment (tamird) [inline sql/update.go]: also there is a TODO at the top of this method that should be removed

Comment (vivekmenezes) [inline sql/update.go]: That's what it should have been. parser.EvalExpr returns a Datum.

File: sql/update.go [hunk 2]
```diff
@@ -90,20 +93,68 @@ func (p *planner) Update(n *parser.Update) (planNode, error) {
 		vals = append(vals, val)
 	}
 
+	if len(cols) != len(vals) {
+		panic("columns not equal updated values")
```
Comment (tamird) [inline sql/update.go]: there's a similar check in `insert.go` but it returns an error rather than panicking.

File: sql/update.go [hunk 3]
```diff
@@ -90,20 +93,68 @@ func (p *planner) Update(n *parser.Update) (planNode, error) {
...
+		// Update the row values.
+		for i, col := range cols {
+			rowVals[colIDtoRowIndex[col.ID]] = vals[i]
+		}
+		newSecondaryIndexEntries, err := encodeSecondaryIndexes(tableDesc.ID, indexes, colIDtoRowIndex, rowVals, primaryIndexKeySuffix)
 		if err != nil {
 			return nil, err
 		}
-		// Update the columns.
+		if len(secondaryIndexEntries) != len(newSecondaryIndexEntries) {
```
Comment (tamird) [inline sql/update.go]: i think this is not a worthwhile check; as long as the `indexes` array passed into `encodeSecondaryIndexes` is unchanged, we're good.

File: sql/update.go [hunk 4]
```diff
@@ -90,20 +93,68 @@ func (p *planner) Update(n *parser.Update) (planNode, error) {
...
+			panic(fmt.Sprintf("unequal number of secondary indexes being updated; %d vs %d",
+				len(secondaryIndexEntries), len(newSecondaryIndexEntries)))
+		}
+		// Update secondary indexes
+		for i, newSecondaryIndexEntry := range newSecondaryIndexEntries {
```
Comment (tamird) [inline sql/update.go]: for symmetry, can you use `i` to index into both arrays?

**Comment by tamird** [comment]
LGTM

File: sql/update.go (hunk 2 above)
Comment (vivekmenezes) [inline sql/update.go]: I've removed this panic. It's really useless given that both slices are created within the same function and equal len(n.Exprs).

File: sql/update.go (hunk 4 above)
Comment (vivekmenezes) [inline sql/update.go]: done

**Comment by vivekmenezes** [comment]
https://github.com/cockroachdb/cockroach/issues/1627

**Comment by petermattis** [comment]
LGTM

**Pull request merged**